### ISSUE TITLE
Key info only added to the following segment

### ISSUE THREAD

**Issue by MoSal** [rutgersc/m3u8-rs#28]
Key info should be added to the following Media Initialization Sections and every segment until the next key.

> ## 4.3.2.4.  EXT-X-KEY
>
>   Media Segments MAY be encrypted.  The EXT-X-KEY tag specifies how to
>   decrypt them.  It applies to **every Media Segment and to every Media
>   Initialization Section declared by an EXT-X-MAP tag that appears
>   between it and the next EXT-X-KEY tag** in the Playlist file with the
>   same KEYFORMAT attribute (or the end of the Playlist file).  Two or
>   more EXT-X-KEY tags with different KEYFORMAT attributes MAY apply to
>   the same Media Segment if they ultimately produce the same decryption
>   key.

**Comment by vagetman** [comment]
The quote states that the key **applies** to every Media Segment. Not that it should be added to every segment. Maybe an example of what you mean can help?

**Comment by MoSal** [comment]
FWIW, I started using a parser I wrote myself which is more suited to my needs. So, this does not affect me anymore.

-----

An example (although it's not necessary):

```
#EXTM3U
#EXT-X-VERSION:4
## Created with OEM(version=1.7.25)
#EXT-X-PLAYLIST-TYPE:VOD
#EXT-X-MEDIA-SEQUENCE:1
#EXT-X-INDEPENDENT-SEGMENTS
#EXT-X-TARGETDURATION:12
#USP-X-TIMESTAMP-MAP:MPEGTS=900000,LOCAL=1970-01-01T00:00:00Z
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-1.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-2.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-3.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-4.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-5.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-6.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-7.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-8.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-9.ts
#EXTINF:6, no desc
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-10.ts
```

> The quote states that the key applies to every Media Segment. Not that it should be added to every segment. Maybe an example of what you mean can help?

What's the difference?

This is arguing semantics, I think. Key information is represented as a field (`pub key: Option<Key>,`) in `MediaSegment`. I think it's reasonable for library users to assume that a `None` value means no encryption, just like how a `None` value of map means no MIS is needed to demux/decrypt.

**Comment by vagetman** [comment]
This is a bit dated issue, but I'd like to get to the bottom of it, if it exists.

> What's the difference?
> 
> This is arguing semantics, I think. Key information is represented as a field (`pub key: Option<Key>,`) in `MediaSegment`. I think it's reasonable for library users to assume that a `None` value means no encryption, just like how a `None` value of map means no MIS is needed to demux/decrypt.

I don't believe I was arguing semantics. As stated in the manifest, the key

```
applies to every Media Segment and to every Media
Initialization Section declared by an EXT-X-MAP tag that appears
between it and the next EXT-X-KEY tag in the Playlist file with the
same KEYFORMAT attribute (or the end of the Playlist file). 
```
Thus, if the key is the same, and it does not need to be `added to the following Media Initialization Sections and every segment until the next key` as you stated.

The manifest you provided is the same as the following 
```
#EXTM3U
#EXT-X-VERSION:4
## Created with OEM(version=1.7.25)
#EXT-X-PLAYLIST-TYPE:VOD
#EXT-X-MEDIA-SEQUENCE:1
#EXT-X-INDEPENDENT-SEGMENTS
#EXT-X-TARGETDURATION:12
#USP-X-TIMESTAMP-MAP:MPEGTS=900000,LOCAL=1970-01-01T00:00:00Z
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-1.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-2.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-3.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-4.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-5.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-6.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-7.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-8.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-9.ts
#EXTINF:6, no desc
#EXT-X-KEY:METHOD=AES-128,URI="https://itvpnp.live.ott.irdeto.com/HLS/GetKey.aspx?CrmId=itvpnp&AccountId=itvpnp&ContentId=2-6292-0647-001_6"
2-6292-0647-001_6_1_VAR003-audio=96523-video=1463000-10.ts
```

Because from the standard the Key is valid until a next key appears in the manifest, such manifest rendering is unnecessary redundant. 

Further, a previous version of the crate used to produce manifest this way, which was addressed in issue #8. The values now are being dropped on every iteration while rendering `MediaPlaylist` from `Vec<MediaPlaylistTag>`

```rust
                SegmentTag::Uri(u) => {
                    next_segment.key = encryption_key.clone();
                    next_segment.map = map.clone();
                    next_segment.uri = u;
                    media_playlist.segments.push(next_segment);
                    next_segment = MediaSegment::empty();
                    encryption_key = None;  // <-- here
                    map = None;             // <-- here
                }
```

Unless there is anything further, I believe the issue should be closed.
